packages/flame/lib/src/collisions/quadtree/broadphase.dart

```diff
@@ -0,0 +1,161 @@
+import 'dart:collection';
+
+import 'package:flame/collisions.dart';
+import 'package:flame/components.dart';
+import 'package:flame/extensions.dart';
+
+typedef ExternalBroadphaseCheck = bool Function(
+  PositionComponent one,
+  PositionComponent another,
+);
+
+typedef ExternalMinDistanceCheck = bool Function(
+  Vector2 activeItemCenter,
+  Vector2 potentialCenter,
+);
+
+/// Performs Quad Tree broadphase check.
+///
+/// See [HasQuadTreeCollisionDetection.initializeCollisionDetection] for a
+/// detailed description of its initialization parameters.
+class QuadTreeBroadphase<T extends Hitbox<T>> extends Broadphase<T> {
+  QuadTreeBroadphase({
+    super.items,
+    required Rect mainBoxSize,
+    required this.broadphaseCheck,
+    required this.minimumDistanceCheck,
+    int maxObjects = 25,
+    int maxLevels = 10,
+  }) : tree = QuadTree<T>(
+          mainBoxSize: mainBoxSize,
+          maxObjects: maxObjects,
+          maxLevels: maxLevels,
+        );
+
+  final QuadTree tree;
+
+  final activeCollisions = HashSet<T>();
+
+  ExternalBroadphaseCheck broadphaseCheck;
+  ExternalMinDistanceCheck minimumDistanceCheck;
+  final _broadphaseCheckCache = <T, Map<T, bool>>{};
+
+  final _cachedCenters = <ShapeHitbox, Vector2>{};
+
+  final _potentials = HashSet<CollisionProspect<T>>();
+  final _potentialsTmp = <List<ShapeHitbox>>[];
+
+  @override
+  HashSet<CollisionProspect<T>> query() {
+    _potentials.clear();
+    _potentialsTmp.clear();
+
+    for (final activeItem in activeCollisions) {
+      final asShapeItem = activeItem as ShapeHitbox;
+
+      if (asShapeItem.isRemoving || asShapeItem.parent == null) {
+        tree.remove(activeItem);
+        continue;
+      }
+
+      final itemCenter = activeItem.aabb.center;
+      final markRemove = <ShapeHitbox>[];
+      final potentiallyCollide = tree.query(activeItem);
+      for (final potential in potentiallyCollide.entries.first.value) {
+        if (potential.collisionType == CollisionType.inactive) {
+          continue;
+        }
+
+        if (_broadphaseCheckCache[activeItem]?[potential] == false) {
+          continue;
+        }
+
+        final asShapePotential = potential as ShapeHitbox;
+
+        if (asShapePotential.isRemoving || asShapePotential.parent == null) {
+          markRemove.add(potential);
+          continue;
+        }
+        if (asShapePotential.parent == asShapeItem.parent &&
+            asShapeItem.parent != null) {
+          continue;
+        }
+
+        Vector2 potentialCenter;
+        if (potential.collisionType == CollisionType.passive) {
+          potentialCenter = _getCenterOfHitbox(asShapePotential);
```

Comment:
Why are the centers only cached when the collion type is passive? :thinking:

Reply:
Because passive collision type usualy means that the component keeps it position, probably whole game. So it worth to cache it's center. But it is pointless for activelly moving object, which would have "active" collisionType most probably.

Reply:
Because passive collision type usualy means that the component keeps it position, probably whole game. So it worth to cache it's center. But it is pointless for activelly moving object, which would have "active" collisionType most probably.

Reply:
That's not what the passive collision type means, say for example bullets, they would usually be passive since there would be alot of them and few things that they should hit, so the player and enemies would be active and the bullets passive to make the collision detection more efficient (this would lead to the bullets not being able to collide with themselves for example).

https://docs.flame-engine.org/1.3.0/flame/collision_detection.html?highlight=collisiontype#collisiontype

Reply:
It mostly depends on desired game logic. It might be a bit different, allowing bullets to collide only if they are fired by players of different fractions.

Anyway, it would be useful for collision system to know, what objects are static and what objects are dynamic - to cache its "center". It turns out to be too expensive to calculate it each time with getter. We can introduce another field for it, if you think that "CollisionType" is not a best choice.

Reply:
Yeah, we can definitely not use `CollisionType` for this since that is already defined how it works.
Can't you just cache all center positions for now?

I don't think this issue should be handled in this PR, but for future reference, In box2d/forge2d what you are describing is called `BodyType`:
https://github.com/flame-engine/forge2d/blob/main/packages/forge2d/lib/src/dynamics/body_type.dart

Reply:
I think the best approach would be like with aabb invalidation: let getter get values always from cached variable an recalculate center and store to cache when transform2d changes.

Reply:
fixed